pixeltable/catalog/table_version.py

```diff
@@ -417,98 +435,110 @@ def init(self) -> None:
-            undo_col._stores_cellmd = False
-
-            # The index is active in this TableVersion provided that:
-            # (i) the TableVersion supports indices (either it's not a snapshot, or it's a replica at
-            #     the head version); and
-            # (ii) the index was created on or before the schema version of this TableVersion; and
-            # (iii) the index was not dropped on or before the schema version of this TableVersion.
-            supports_idxs = self.effective_version is None or (
-                self.tbl_md.is_replica and self.effective_version == self.tbl_md.current_version
+            stores_cellmd: bool | None = None  # None: determined by the column properties (in the Column c'tor)
+            sa_col_type: sql.sqltypes.TypeEngine | None = None
+            if col_md.id in val_col_idxs:
+                idx = val_col_idxs[col_md.id]
+                # for index value columns, the index gets to override the default
+                stores_cellmd = idx.records_value_errors()
+                sa_col_type = idx.get_index_sa_type(col_type)
+            elif col_md.id in undo_col_idxs:
+                idx = undo_col_idxs[col_md.id]
+                # for index undo columns, we never store cellmd
+                stores_cellmd = False
+                sa_col_type = idx.get_index_sa_type(col_type)
+
+            col = Column(
+                col_id=col_md.id,
+                name=schema_col_md.name if schema_col_md is not None else None,
+                col_type=col_type,
+                is_pk=col_md.is_pk,
+                is_iterator_col=self.is_component_view and col_md.id < self.num_iterator_cols + 1,
+                stored=col_md.stored,
+                media_validation=media_val,
+                sa_col_type=sa_col_type,
+                schema_version_add=col_md.schema_version_add,
+                schema_version_drop=col_md.schema_version_drop,
+                stores_cellmd=stores_cellmd,
+                value_expr_dict=col_md.value_expr,
+                tbl_handle=self.handle,
+                destination=col_md.destination,
             )
-            if (
-                supports_idxs
-                and md.schema_version_add <= self.schema_version
+
+            self.cols.append(col)
+            # populate lookup structures before Expr.from_dict()
+            if col_md.schema_version_add <= self.schema_version and (
+                col_md.schema_version_drop is None or col_md.schema_version_drop > self.schema_version
+            ):
+                if col.name is not None:
+                    self.cols_by_name[col.name] = col
+                self.cols_by_id[col.id] = col
+
+        if self.supports_idxs:
+            # create IndexInfo for indices visible in current_version
+            for md in (
+                md
+                for md in self.tbl_md.index_md.values()
+                if md.schema_version_add <= self.schema_version
```

Comment:
Add a code comment explaining what this relatively complicated condition represents. (There was one in there before.)